src/models/dirac_mixture_single_publisher_model.py

```diff
@@ -0,0 +1,245 @@
+
+    @staticmethod
+    def solve_optimal_weights(
+        observed_arr: np.ndarray, component_arrs: List, distance: Callable
+    ) -> np.ndarray:
+        """Convex optimization of weights on different components.
+
+        Args:
+            observed_arr: Any 1d or 2d arry.
+            component_arrs: A list of arrays that are of the same dimension
+                of observed_arr.
+            distance: A function on two arrays of the same dimension. Calculates
+                a certain distance between the two arrays.
+
+        Returns:
+            A weight vector such that the weighted sum of the arrays in
+            component_arrs is closest to observed_arr with respect to the given
+            distance.
+        """
+        ws = cp.Variable(len(component_arrs))
+        pred = cp.sum([cp.multiply(w, arr) for w, arr in zip(ws, component_arrs)])
+        problem = cp.Problem(
+            objective=cp.Minimize(distance(observed_arr, pred)),
+            constraints=[ws >= 0, cp.sum(ws) == 1],
+        )
+        # cvxpy is an interface of multiple convex optimization solver packages.
+        # Among them, the open source packages are ECOS
+        # (https://github.com/embotech/ecos/wiki), SCS
+        # (https://www.cvxgrp.org/scs/api/index.html) and OSQP, where ECOS and
+        # SCS are able to handle our problem. So, we try ECOS first and then SCS.
+        # Note: ECOS is available in C.  SCS available in C/C++.  But they are
+        # both unavailable in Java.
+        try:
+            problem.solve()  # The default solver is ECOS
+        except cp.SolverError:
+            problem.solve(solver=cp.SCS)
+        return ws.value
+
+    def fit(self):
+        """Fits a univariate mixed Poisson distribution."""
+        if self.fitted:
+            return
+        self.component_pmfs = [
+            self.truncated_poisson_pmf_vec(component, self.max_freq)
+            for component in self.components
+        ]
+        self.ws = self.solve_optimal_weights(
+            self.observed_pmf,
+            self.component_pmfs,
+            self.cross_entropy,
+        )
+        self.fitted = True
+
+    def predict(
+        self, scaling_factor: float, customized_max_freq: int = None
+    ) -> np.ndarray:
+        """Predict the frequency histogram when scaling the number of impressions.
+
+        The Dirac mixture model assumes that when scaling the number of
+        impressions, different Poisson components are scaled accordingly,
+        and that the pmf is shifted to the weighted sum of these scaled
+        Poisson distributions with the same weights.
+
+        Explicitly, if for the actual campaign its frequency distribution
+        is fitted as
+        sum_j w_j Poisson(lambda_j),
+        Then for a hypothetical campaign that has alpha times the number
+        of impressions of the actual campaign, its frequency distribution
+        is predicted as
+        sum_j w_j Poisson(alpha * lambda_j).
```

Comment:
Ok I understand this comment and I agree with it.

However, I have a different suggestion for the choice of normalization. I think getting the gross rating point of the campaign or (impression/person) and fitting to a normalized distribution might be more stable. I will test this and get back to you. We can also discuss in a call.

Reply:
Thanks for the suggestion.  I haven't understood your point, and look forward to a discussion.  FYI, Matthew and I were planning to apply different models to different demo buckets, so GRP should be equivalent to #impressions.  Not sure if it's relevant to your comment!

Reply:
I need to write this down and then let's talk in a zoom call instead. For now feel free to resolve the conversation. Thanks!